Bluetooth: Audio: Media - pass structs by reference instead of by value

asbjornsabo opened the pull request:
Change the MCC (media control client) and media proxy APIs to pass structs by reference instead of by value.
- optimization - less to pass around
- consistency - this is how structs are usually passed in comparable files

Note that these APIs are marked as experimental, so API changes should be OK.

Also changes internal functions in the media player, the media client test, and the media controller test to pass structs by reference instead of by value.

closes https://github.com/zephyrproject-rtos/zephyr/issues/42096

asbjornsabo committed:
Bluetooth: Audio: MCC - pass structs by reference

Changes the media control client API to pass structs by reference
instead of by value.

Also change internals of test to pass struct by reference.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Media control - pass structs by reference

Changes the media_proxy API to pass structs by reference instead of by
value.

Also pass structs by reference in internal functions in media player
and in media controller test.

Add pointer checks to the public API implementation where
user/application code provides structs by reference.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

jhedberg commented:
Does this also consider whether the function needs to modify the parameter, and if not use `const struct foo *`?

asbjornsabo commented:
> Does this also consider whether the function needs to modify the parameter, and if not use `const struct foo *`?

No, that was forgotten.  Will fix.

Thalley commented on subsys/bluetooth/audio/mpl.c:
```diff
@@ -1740,68 +1740,68 @@ void playing_state_command_handler(struct mpl_cmd command,
 	default:
-		BT_DBG("Invalid command: %d", command.opcode);
-		ntf.result_code = BT_MCS_OPC_NTF_NOT_SUPPORTED;
+		BT_DBG("Invalid command: %d", command->opcode);
+		ntf->result_code = BT_MCS_OPC_NTF_NOT_SUPPORTED;
```
Unrelated to the changes in this PR but it looks like we always override `ntf->result_code` before calling `media_proxy_pl_command_cb`. If that is the case, why do we even use a `mpl_cmd_ntf` as a parameter instead of just the `requested_opcode`, and then just create the `mpl_cmd_ntf` in the command handlers?

asbjornsabo replied:
It is not over-ridden, it is set.
But yes, could do as you suggest.  It adds one more variable (opcode + `mpl_cmd_ntf` instead of just the last one), but maybe it looks cleaner.
I believe the thought was that the notification struct must be filled in with values from two places in the code, so it was just as easy to send it along.
No strong opinion on this one.

Thalley replied:
> It adds one more variable (opcode + mpl_cmd_ntf instead of just the last one),

If we send the opcode by parameter, we don't need the `mpl_cmd_ntf` argument at all, do we? 

mpl_cmd_ntf contains the opcode and a result code; if we get the opcode from the caller, and the result code from this function, this function can just allocate the `mpl_cmd_ntf` struct, right? 

It is outside the scope of this PR, so do with it as you will :)

Thalley commented on subsys/bluetooth/audio/mpl.c:
```diff
@@ -2607,38 +2607,38 @@ uint32_t get_commands_supported(void)
 			break;
 		}
-		memcpy(&sci.param, &search.search[index], sci.len - 1);
+		memcpy(&sci.param, &search->search[index], sci.len - 1);
```
```suggestion
		(void)memcpy(&sci.param, &search->search[index], sci.len - 1);
```

asbjornsabo replied:
Fixed

Thalley reviewed (changes requested):
I think we need to consider using `CHECKIF` for the new pointers, and implement it or at least make a conscious choice to not check them. 

As Johan also mentioned, most of these can probably be marked as `const` which I am in favor of.

asbjornsabo committed:
Bluetooth: Audio: Remove zero-ing of variable

Commit dc812539b81a ("Bluetooth: MPL: Fix use of uninitialized
command.param value") fixed the use of a potentially un-initialized
variable, by
- splitting a log statement, so that the variable is only
  logged (=used) if in use (and therefore hopefully also initialized)
- setting the variable to zero if not in use

The second part is not required, as all other code in the module
checks whether the variable is in use before using it.  And this
second part also prevents the parent struct from being made 'const'.

This commits removes the zeroing of the variable, so that the struct
can be made 'const'.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Media - search to not modify parameter

Update the search parser to not modify the supplied parameter.
This will allow the parameter to be made 'const'.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented:
Updated with 'const' keywords where appropriate (=everywhere).
Added a couple of small commits to allow  to make all structs const.

asbjornsabo committed:
Bluetooth: Audio: Add void() cast to memcpy

Cast memcpy to void().

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -2158,11 +2158,11 @@ int bt_mcc_send_cmd(struct bt_conn *conn, struct mpl_cmd cmd)
+	if (cmd->use_param) {
+		length += sizeof(cmd->param);
+		memcpy(&cur_mcs_inst->write_buf[sizeof(cmd->opcode)], &cmd->param,
+		       sizeof(cmd->param));
```
Fixed

asbjornsabo commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -2209,7 +2209,7 @@ int bt_mcc_read_opcodes_supported(struct bt_conn *conn)
 
 #ifdef CONFIG_BT_MCC_OTS
-int bt_mcc_send_search(struct bt_conn *conn, struct mpl_search search)
+int bt_mcc_send_search(struct bt_conn *conn, struct mpl_search *search)
```
Proposal:
- Do not check in the callbacks - pointers there will be provided by the stack code, which we choose to trust
- Do check in the media proxy calls - values here will be provided by the application/user
- Do not check in the media control client calls - the plan is to use these through the media proxy (only), so checks will eventually double up.

asbjornsabo replied:
Did add CHECKIF to the places in the media_proxy API where user/application code supplies structs by reference.
Did not add checks to the callbacks, or to MCC, as described above.


> Does this also consider whether the function needs to modify the parameter, and if not use `const struct foo *`?

@jhedberg Now `const` everywhere.

Thalley commented on subsys/bluetooth/audio/media_proxy.c:
```diff
@@ -197,7 +197,7 @@ uint8_t media_proxy_sctrl_get_media_state(void)
 }
 
-void media_proxy_sctrl_send_command(struct mpl_cmd cmd)
+void media_proxy_sctrl_send_command(struct mpl_cmd *cmd)
```
Could `cmd` be `const` here as well? Ideally the command handler should not need to to modify the command.

asbjornsabo commented:
Added CHECKIF to the places where user/application code submits structs by reference.
Added `const` structs to a couple of more internal calls.

mbolivar-nordic commented:
@asbjornsabo please rebase

asbjornsabo commented:
@mbolivar-nordic Rebased, with no additional changes.

asbjornsabo commented:
Fix commit message that did not pass compliance - use correct format for reference to other commit.
